File: QuasarCode_Library_Maths/Units/UnitBase.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace QuasarCode.Library.Maths.Units
{
    public abstract class UnitBase : IUnit
    {
        public abstract FundamentalUnitPowerPair[] GetFundamentalUnitPairs();
        public abstract UnitPowerPair[] GetUnitPairs();
        public abstract IUnit Pow(int p);
        public abstract object Clone();

        public IUnit Mult(IUnit u)
        {
            List<UnitPowerPair>  items = new List<UnitPowerPair>(this.GetUnitPairs());
            items.AddRange(u.GetUnitPairs());
            return CompoundUnit.NewCompoundUnit(items.ToArray());
        }

        public IUnit Div(IUnit u)
        {
            List<UnitPowerPair> items = new List<UnitPowerPair>(this.GetUnitPairs());

            foreach (UnitPowerPair pair in u.GetUnitPairs())
            {
                UnitPowerPair newPair = new UnitPowerPair { Unit = pair.Unit, Power = -pair.Power };
                items.Add(newPair);
            }

            return CompoundUnit.NewCompoundUnit(items.ToArray());
        }

        public static IUnit operator *(UnitBase a, IUnit b)
        {
            return a.Mult(b);
        }

        public static IUnit operator /(UnitBase a, IUnit b)
        {
            return a.Div(b);
        }

        public override string ToString()
        {
            string result = "";

            foreach (FundamentalUnitPowerPair pair in GetFundamentalUnitPairs())
            {
                result += pair.Unit.ToString();
                foreach (char character in pair.Power.ToString())
                {
                    if (character == '-')
                    {
                        result += QuasarCode.Library.Tools.StringLiterals.Superscript_Minus;
                    }
                    else
                    {
                        result += Tools.StringLiterals.SuperscriptInt[Convert.ToInt32(character)];
                    }
                }

                result += " ";
            }

            return result.Trim();
        }
    }
}

File: QuasarCode_Python/QuasarCode/Maths/Units/_UnitBase.py
from QuasarCode.Maths.Units._IUnit import IUnit
from QuasarCode.Tools import StringLiterals

class UnitBase(IUnit):
    def Mult(self, u):
        return CompoundUnit.NewCompoundUnit(self.GetUnitPairs().extend(u.GetUnitPairs()))

    def Mult(self, u):
        items = self.GetUnitPairs()

        for pair in u.GetUnitPairs():
            items.append(UnitPowerPair(Unit = pair.Unit, Power = -pair.Power))

        return CompoundUnit.NewCompoundUnit(items)

    def __mul__(self, other):
        return self.Mult(other)

    def __div__(self, other):
        return self.Div(other)

    def ToString(self):
        result = ""

        for pair in self.GetFundamentalUnitPairs():
            result += pair.Unit.ToString()
            for character in pair.Power.ToString():
                if character == "-":
                    result += StringLiterals.Superscript_Minus
                else:
                    result += StringLiterals.superscriptFromNormal(character)

            result += " ";

        return result.rstrip();